File: python/graphs/prim.py
## Taken from https://github.com/CianLR/judge-solutions/blob/master/kattis/islandhopping.py
## Tested against kattis:islandhopping, kattis:cats.

import heapq

def complex_dist(a, b):
    import cmath
    return cmath.polar(a - b)[0]

# Takes a list of points of the form complex(x, y).
# Can be modified to take edge weights if the line marked #metric is changed.
# Runs in time proportional to the number of vertices.
def prims(N, points):
    cost = 0
    pq = [(0, 0)]
    in_tree = [False] * N
    tree_dist = [1000000] * N
    tree_size = 0
    while tree_size < N and pq:
        d, u = heapq.heappop(pq)
        if in_tree[u]:
            continue
        in_tree[u] = True
        cost += d
        tree_size += 1
        for v in range(N):
            if u == v or in_tree[v]:
                continue
            dist = complex_dist(points[u], points[v]) #metric
            if dist > tree_dist[v]:
                continue
            tree_dist[v] = dist
            heapq.heappush(pq, (dist, v))
    return cost

File: cpp/graphs/prim.cc
# Taken from https://github.com/CianLR/judge-solutions/blob/d273867c04813a161ec888aecf47e5d6c2dc916a/kattis/freckles_prims.cc
# Tested against Kattis:freckles

#include <math.h>
#include <functional>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

struct Point {
  Point(double x_, double y_) : x(x_), y(y_), in_tree(false) {}
  double x;
  double y;
  bool in_tree;
};

double Dist(const Point &a, const Point &b) {
  return sqrt(pow(a.x - b.x, 2) + pow(a.y - b.y, 2));
}

double PrimsMST(int N, const vector<Point> &pts) {
  typedef pair<double, Point *> CostPoint;
  priority_queue<CostPoint, vector<CostPoint>, greater<CostPoint>> pq;
  pq.emplace(0, &pts[0]);
  double cost = 0;
  int total_nodes = 0;
  while (total_nodes < N) {
    if (pq.top().second->in_tree) {
      pq.pop();
      continue;
    }
    cost += pq.top().first;
    ++total_nodes;
    Point *p = pq.top().second;
    pq.pop();
    p->in_tree = true;
    for (int i = 0; i < N; ++i) {
      if (!pts[i].in_tree)
        pq.emplace(Dist(*p, pts[i]), &pts[i]);
    }
  }
  return cost;
}

